### PR TITLE
Switch from Lager to the new Erlang Logger API for logging

### PR DESCRIPTION
**Pull request by dumbbell**
Switch from Lager to the new Erlang Logger API for logging

The configuration remains the same for the end-user. The only exception is the log root directory: it is now set through the `log_root` application env. variable in `rabbit`. People using the Cuttlefish-based configuration file are not affected by this exception.

The main change is how the logging facility is configured. It now happens in `rabbit_prelaunch_logging`. The `rabbit_lager` module is removed.

The supported outputs remain the same: the console, text files, the `amq.rabbitmq.log` exchange and syslog.

The message text format slightly changed: the timestamp is more precise (now to the microsecond) and the level can be abbreviated to always be 4-character long to align all messages and improve readability. Here is an example:

    2021-03-03 10:22:30.377392+01:00 [dbug] <0.229.0> == Prelaunch DONE ==
    2021-03-03 10:22:30.377860+01:00 [info] <0.229.0>
    2021-03-03 10:22:30.377860+01:00 [info] <0.229.0>  Starting RabbitMQ 3.8.10+115.g071f3fb on Erlang 23.2.5
    2021-03-03 10:22:30.377860+01:00 [info] <0.229.0>  Licensed under the MPL 2.0. Website: https://rabbitmq.com

The example above also shows that multiline messages are supported and each line is prepended with the same prefix (the timestamp, the level and the Erlang process PID).

JSON is also supported as a message format and now for any outputs.  Indeed, it is possible to use it with e.g. syslog or the exchange. Here is an example of a JSON-formatted message sent to syslog:

    Mar  3 11:23:06 localhost rabbitmq-server[27908] <0.229.0> - {"time":"2021-03-03T11:23:06.998466+01:00","level":"notice","msg":"Logging: configured log handlers are now ACTIVE","meta":{"domain":"rabbitmq.prelaunch","file":"src/rabbit_prelaunch_logging.erl","gl":"<0.228.0>","line":311,"mfa":["rabbit_prelaunch_logging","configure_logger",1],"pid":"<0.229.0>"}}

For quick testing, the values accepted by the `$RABBITMQ_LOGS` environment variables were extended:
  * `-` still means stdout
  * `-stderr` means stderr
  * `syslog:` means syslog on localhost
  * `exchange:` means logging to `amq.rabbitmq.log`

`$RABBITMQ_LOG` was also extended. It now accepts a `+json` modifier (in addition to the existing `+color` one). With that modifier, messages are formatted as JSON intead of plain text.

The `rabbitmqctl rotate_logs` command is deprecated. The reason is Logger does not expose a function to force log rotation. However, it will detect when a file was rotated by an external tool.

From a developer point of view, the old `rabbit_log*` API remains supported, though it is now deprecated. It is implemented as regular modules: there is no `parse_transform` involved anymore.

In the code, it is recommended to use the new Logger macros. For instance, `?LOG_INFO(Format, Args)`. If possible, messages should be augmented with some metadata. For instance (note the map after the message):

    ?LOG_NOTICE("Logging: switching to configured handler(s); following "
                "messages may not be visible in this log output",
                #{domain => ?RMQLOG_DOMAIN_PRELAUNCH}),

Domains in Erlang Logger parlance are the way to categorize messages.  Some predefined domains, matching previous categories, are currently defined in `rabbit_common/include/logging.hrl` or headers in the relevant plugins for plugin-specific categories.

At this point, very few messages have been converted from the old `rabbit_log*` API to the new macros. It can be done gradually when working on a particular module or logging.

The Erlang builtin console/file handler, `logger_std_h`, has been forked because it lacks date-based file rotation. The configuration of date-based rotation is identical to Lager. Once the dust has settled for this feature, the goal is to submit it upstream for inclusion in Erlang.  The forked module is calld `rabbit_logger_std_h` and is based `logger_std_h` in Erlang 23.0.

### CODE REVIEW

**Comment by michaelklishin** [comment]
Why was `rabbitmqctl rotate_logs` removed? If we don't have it, it would be much harder to get this into 3.8.x one day, which means supporting Erlang 24 there potentially would be impossible.

**Comment by michaelklishin** [comment]
There is a subtle change to severity formatting in the logs. Note how it uses `[dbug]` instead of `[debug]` which most tools use:

```
2021-03-05 01:39:36.577765+03:00 [dbug] <0.6137.0> User 'guest' authenticated successfully by backend rabbit_auth_backend_internal
```

**Comment by dumbbell** [comment]
> Why was `rabbitmqctl rotate_logs` removed? If we don't have it, it would be much harder to get this into 3.8.x one day, which means supporting Erlang 24 there potentially would be impossible.

Erlang Logger does not offer a way to force log rotation. However it will detect correctly when a file was rotated by an external tool. We can restore `rotate_logs` if we backport but it will be a no-op and should return an error.

**Comment by dumbbell** [comment]
> There is a subtle change to severity formatting in the logs. Note how it uses `[dbug]` instead of `[debug]` which most tools use:
> 
> ```
> 2021-03-05 01:39:36.577765+03:00 [dbug] <0.6137.0> User 'guest' authenticated successfully by backend rabbit_auth_backend_internal
> ```

That's me who wanted aligned messages. It made the logs much more readable during development. I plan to make this configurable (it is already in the code), but I don't know yet how to expose the new settings. I would like to have more feedback from the Platform team because they will need the JSON structure to be configurable as well probably. Once I have more input, I can figure out a good way to expose the configuration.

**Comment by dumbbell** [comment]
Thank you for the patch for Erlang 24. I squashed it in the first commit.

**Comment by michaelklishin** [comment]
@dumbbell I suspect there is a lot of systems that depend on `rabbitmqctl rotate_logs`. So I'd reintroduce it and make it print a warning and do nothing, or find a way to perform rotation. Windows users specifically do not have a lot of options when it comes to log rotation, in fact, we see that some Windows users learn about log rotation from RabbitMQ :(

**Comment by michaelklishin** [comment]
Speaking of Windows, a bit of research yields [this comment](https://github.com/influxdata/telegraf/issues/3393#issuecomment-378587457) which suggests that external log rotation is indeed something that Windows users cannot use and/or frown upon.

**Comment by michaelklishin** [comment]
I've noticed that we now have excessive logging around supervisor startup, e.g.

```
2021-03-05 16:05:40.116547+03:00 [info] <0.1027.0> Supervisor {<0.1027.0>,ranch_acceptors_sup}: child {acceptor,<0.1027.0>,1} started (<0.1030.0>): {ranch_acceptor,start_link,[{sslsocket,nil,{#Port<0.42>,{config,#{use_ticket => undefined,crl_cache => {ssl_crl_cache,{internal,[]}},eccs => {elliptic_curves,[{1,3,132,0,39},{1,3,132,0,38},{1,3,132,0,35},{1,3,36,3,3,2,8,1,1,13},{1,3,132,0,36},{1,3,132,0,37},{1,3,36,3,3,2,8,1,1,11},{1,3,132,0,34},{1,3,132,0,16},{1,3,132,0,17},{1,3,36,3,3,2,8,1,1,7},{1,3,132,0,10},{1,2,840,10045,3,1,7},{1,3,132,0,3},{1,3,132,0,26},{1,3,132,0,27},{1,3,132,0,32},{1,3,132,0,33},{1,3,132,0,24},{1,3,132,0,25},{1,3,132,0,31},{1,2,840,10045,3,1,1},{1,3,132,0,1},{1,3,132,0,2},{1,3,132,0,15},{1,3,132,0,9},{1,3,132,0,8},{1,3,132,0,30}]},verify => verify_none,secure_renegotiate => true,fallback => undefined,erl_dist => false,dhfile => undefined,ocsp_responder_certs => [],ocsp_nonce => true,customize_hostname_check => [],keyfile => <<"/Users/antares/Development/Opensource/tls-gen.git/basic/result/server_key.pem">>,partial_chain => #Fun<ssl.11.3360443>,server_name_indication => undefined,max_fragment_length => undefined,dh => undefined,signature_algs_cert => undefined,key => undefined,alpn_preferred_protocols => undefined,sni_fun => undefined,reuse_session => #Fun<ssl.13.3360443>,verify_fun => {#Fun<ssl.12.3360443>,[]},log_level => notice,cookie => true,next_protocol_selector => undefined,handshake => full,hibernate_after => infinity,key_update_at => 388736063997,ciphers => [<<19,2>>,<<19,1>>,<<19,3>>,<<19,4>>,<<19,5>>,<<"À,">>,<<"À0">>,<<"À­">>,<<"À¯">>,<<"À$">>,<<"À(">>,<<"Ì©">>,<<"Ì¨">>,<<"À+">>,<<"À/">>,<<"À¬">>,<<"À®">>,<<"À.">>,<<"À2">>,<<"À&">>,<<"À*">>,<<"À-">>,<<"À1">>,<<"À#">>,<<"À'">>,<<"À%">>,<<"À)">>,<<0,159>>,<<0,163>>,<<0,107>>,<<0,106>>,<<0,158>>,<<0,162>>,<<"Ìª">>,<<0,103>>,<<0,64>>,<<"À\n">>,<<192,20>>,<<0,57>>,<<0,56>>,<<192,5>>,<<192,15>>,<<"À\t">>,<<192,19>>,<<0,51>>,<<0,50>>,<<192,4>>,<<192,14>>],cacerts => undefined,depth => 10,alpn_advertised_protocols => undefined,psk_identity => undefined,honor_ecc_order => false,cert => undefined,sni_hosts => [],early_data => undefined,max_handshake_size => 262144,beast_mitigation => one_n_minus_one,honor_cipher_order => false,renegotiate_at => 268435456,fail_if_no_peer_cert => false,srp_identity => undefined,client_renegotiation => true,anti_replay => undefined,crl_check => false,signature_algs => [ecdsa_secp521r1_sha512,ecdsa_secp384r1_sha384,ecdsa_secp256r1_sha256,rsa_pss_pss_sha512,rsa_pss_pss_sha384,rsa_pss_pss_sha256,rsa_pss_rsae_sha512,rsa_pss_rsae_sha384,rsa_pss_rsae_sha256,rsa_pkcs1_sha512,rsa_pkcs1_sha384,rsa_pkcs1_sha256,ecdsa_sha1,rsa_pkcs1_sha1,{sha512,ecdsa},{sha512,rsa},{sha384,ecdsa},{sha384,rsa},{sha256,ecdsa},{sha256,rsa},{sha224,ecdsa},{sha224,rsa},{sha,ecdsa},{sha,rsa},{sha,dsa}],user_lookup_fun => undefined,cacertfile => <<"/Users/antares/Development/Opensource/tls-gen.git/basic/result/ca_certificate.pem">>,protocol => tls,middlebox_comp_mode => true,certfile => <<"/Users/antares/Development/Opensource/tls-gen.git/basic/result/server_certificate.pem">>,password => [],padding_check => true,supported_groups => {supported_groups,[x25519,x448,secp256r1,secp384r1]},reuse_sessions => true,versions => [{3,4},{3,3},{3,2},{3,1}],keep_secrets => false,session_tickets => disabled,ocsp_stapling => false,next_protocols_advertised => undefined},[{reuseaddr,true},{send_timeout_close,true},{send_timeout,30000},{nodelay,true},{backlog,1024},{port,15671},{packet_size,0},{packet,0},{header,0},{active,false},{mode,binary}],[{mode,binary},{active,false},{packet,raw},{packet_size,0},{header,0}],[{option_tracker,<0.1028.0>},{session_tickets_tracker,disabled},{session_id_tracker,<0.1029.0>}],undefined,[{reuseaddr,true},{send_timeout_close,true},{send_timeout,30000},{nodelay,true},{backlog,1024},{port,15671},{packet_size,0},{packet,0},{header,0},{active,false},{mode,binary}],{gen_tcp,tcp,tcp_closed,tcp_error,tcp_passive},tls_gen_connection}}},ranch_ssl,error_logger,<0.1026.0>]}
```

at `info` level. This can be ill received by some. Note that it logs the private key password.

File: deps/rabbit/apps/rabbitmq_prelaunch/src/rabbit_logger_std_h.erl [hunk 1]
```diff
@@ -51,7 +51,7 @@
 -export([log/2, adding_handler/1, removing_handler/1, changing_config/3,
          filter_config/1]).
 
--define(DEFAULT_CALL_TIMEOUT, 5000).
+-define(DEFAULT_CALL_TIMEOUT, 15000).
```
Comment (dumbbell) [inline deps/rabbit/apps/rabbitmq_prelaunch/src/rabbit_logger_std_h.erl:54]: This module is a fork of `logger_std_h` to add and stabilise date-based file rotation. The goal is to submit the change upstream once we are happy with it.

Therefore I would like this module to remain as close as possible to the original source and thus I'm going to revert this change.

**Comment by michaelklishin** [comment]
We have lost support for `log.default.level` in the process:

``` ini
log.console = false
log.default.level = warning
log.connection.level = error
```

doesn't have the desired effect (the node logs at `debug` level)

**Comment by dumbbell** [comment]
> We have lost support for `log.default.level` in the process:
> 
> ```ini
> log.console = false
> log.default.level = warning
> log.connection.level = error
> ```
> 
> doesn't have the desired effect (the node logs at `debug` level)

We discussed this outside of GitHub: the reason is that when the node is started using `gmake run-broker`, we override automatically the log level to `debug`. That's why the configuration file is ignored.

**Comment by dumbbell** [comment]
The latest changes (squashed into the main commit and force-pushed) include:
* the update to [syslog 4.0.0](https://github.com/schlagert/syslog/tree/4.0.0#400)
* the renaming of `unit_log_config_SUITE` to `logging_SUITE`
* a new test case for JSON formatting
* fixes after testing the patch on Microsoft Windows 2019
